proposals/p1964.md

```diff
@@ -0,0 +1,356 @@
+-   Character literals support some back-slash (`\`) escape sequences, including
+    `\t`, `\n`, `\r`, `\"`, `\'`, `\\`, `\0`, and `\u{HHHH...}`. See
+    [String Literals: Escape sequence](https://github.com/carbon-language/carbon-lang/blob/trunk/proposals/p0199.md#escape-sequences).
+-   Character literals implicitly convert to Unicode strings, even though the
+    numeric values of the encoding may change.
+
+We will not support:
+
+-   character literals that don't contain exactly one Unicode code point;
+-   multi-line literals;
+-   "raw" literals (using #'x'#);
+-   `\x` escape sequences;
+-   character literals with a single quote (`'`) or back-slash (`\`), except as
+    part of an escape sequence;
+-   empty character literals (`''`);
+-   ASCII control codes (0...31), including whitespace characters other than
+    word space (tab, line feed, carriage return, form feed, and vertical tab),
+    except when specified with an escape sequence.
+
+### Types
+
+We will have the types `Char8`, `Char16`, and `Char32` representing code units
+in UTF-8, UTF-16, and UTF-32, but we will not support all code units, but only
+those which map directly to the complete value of a code point. However,
+character literals will use their own types distinct from these:
+
+-   We will support value preserving implicit conversions from character
+    literals to code point or code unit types. In particular, a character
+    literal converts to a `Char8` UTF-8 code unit if it is less than or equal to
+    0x7F, and `Char16` UTF-16 code unit if it is less than or equal to 0xFFFF.
+-   Conversions from string or character literals to a non-value-preserving
+    encoding must be explicit.
+-   Conversions from string literals to Unicode strings are implicit, even
+    though the numeric values of the encoding may change.
+
+We can see whether the particular literal is represented in the variable's type
+by only looking at the types.
+
+```
+let allowed: Char8 = 'a';
+```
+
+The above is allowed because the type of `'a'` is the character literal
+consisting of the single Unicode code point 97, which can be converted to
+`Char8` since 97 is less than or equal to 0x7F.
+
+```
+let error1: Char8 = '😃';
+let error2: Char8 = 'AB';
+```
+
+However these should produce errors. The type of `'😃'` is the character literal
+consisting of the single Unicode code point `0x1F603`, which is greater than
+0x7F. The type of `'AB'` is a character literal that is a sequence of two
+Unicode code points, which is not valid.
+
+Literals `'\n'` and `'\u{A}'` represent the same character and so have the same
+type. However, explicitly converting this character literal to another character
+set might result in a character with a different value, but that still
+represents the newline character.
+
+```
+// 0x15 is the new line character in EBCDIC.
+Assert('\n' as EBCDICChar == EBCDICChar.Make(0x15));
+Assert('\u{A}' as EBCDICChar != EBCDICChar.Make(0x0A));
+```
```

Comment:
I feel like it would be good to document how conversion from character literals to encoding-specific types ([Shift JIS](https://en.wikipedia.org/wiki/Shift_JIS), etc.) is intended to work. To me it seems good if the resulting character matches what it looks like in the source code, which means doing a conversion from Unicode.

Reply:
I agree, but I think we need to have a pretty complete view of encoding and decoding semantics to examine this effectively. For example, I would want to know how things are the same or different to doing an analogous operation on *string* literals, as well as strings, etc.

I'm mostly trying to restrict the scope here to the core things we really *need* in order to make the character literal part of the language work.

My hope is that the rest can be added as a more pure and focused standard library proposal.